feat(docs): Add Move view functions docs and examples (#11931)
iotaledger/iota: docs/content/developer/move/explanations/view-functions.mdx

---
description: 'What view functions are in Move on IOTA, the read-only guarantees they provide, and how those guarantees are enforced and recorded on-chain.'
tags:
  - explanation
  - move-sc
  - move-vm
---

# View Functions

> A _view function_ is a Move function annotated with the `#[view]` attribute. The attribute marks the function as **read-only**: it may inspect on-chain state but cannot modify it. IOTA enforces this guarantee at both compile time and verifier time, and records it on-chain, so that off-chain consumers can rely on a function being a view without trusting its author.

Reviewer: ```suggestion
A _ Move View Function_ is a standardized interface for enabling ergonomic and application-specific queries to on-chain state. Move view functions are developer defined on-chain read API that can be easily queried off-chain without requiring transaction signing or state mutation. In a Move module, the view function is annotated with the `#[view]` attribute.  
The attribute marks the function as **read-only**: it may inspect on-chain state but cannot modify it. IOTA enforces this guarantee at both compile time and verifier time, and records it on-chain, so that off-chain consumers can rely on a function being a view without trusting its author.
```

Reviewer: Took from iip-5

## Why view functions exist

> Reading data from a Move package has always been possible by simulating a call, but nothing about a function's declaration told a client whether calling it was safe to treat as a pure read. Any function could mutate objects, transfer assets, or emit events as a side effect of being called.

Reviewer: ```suggestion
Reading data from a Move package has always been possible by simulating a call, but nothing about a function's declaration told a client whether calling it was safe to treat as a pure read. Moreover, one must write a significant amount of custom client-side logic to inspect on-chain data. This often involves multiple layers of fetching object IDs, deserializing raw bytes, and navigating dynamic fields just to reach an information stored within a Move object. 
```

> The `#[view]` attribute closes that gap. It turns "this function only reads state" from an informal convention into a **verified property**:

Reviewer: ```suggestion
The Move View Function closes that gap:
```

> - **Package authors** express intent directly in the source, and the compiler holds them to it.
> - **Client and SDK developers** can discover a package's view functions and call them knowing the call has no side effects.
> - **Tooling** (explorers, indexers, wallets) can surface view functions as a safe, queryable read API for a package.

Reviewer: ```suggestion
- **Package authors** express intent directly in the source, and the compiler holds them to it.
- **Client and SDK developers** can discover a package's view functions and call them knowing the call has no side effects.
- **Tooling** (explorers, indexers, wallets) can surface view functions as a safe, queryable read API for a package.
- **Extends current RPC** by enabling easy access to data structures using custom fields or dynamic fields, e.g., Bag and Table.
- ** Returned results** are resolved, i.e., the Move types are deserialized.
```

> The motivation, design, and on-chain encoding are specified in [IIP-0011](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0011/IIP-0011.md).

Reviewer: ```suggestion
The motivation, design, and on-chain encoding are specified in [IIP-0005](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0005/iip-0005.md) and [IIP-0011](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0011/IIP-0011.md).
```

## What is a `#[view]` function

A `#[view]` function cannot take a mutable reference (through which it could mutate state), cannot consume objects by value (which would let it destroy or transfer them), and must return a value. For the exact rules, see the [view functions reference](../references/view-functions.mdx).

> 

Reviewer: ```suggestion

This is an example of view function that reads a `Bid` value from an `Auction` object:
```move file=<rootDir>/examples/move/view_functions/sources/auction.move#L83-L97
```

```

---
description: 'What view functions are in Move on IOTA, the read-only guarantees they provide, and how those guarantees are enforced and recorded on-chain.'
tags:
  - explanation
  - move-sc
  - move-vm
---

# View Functions

A _ Move View Function_ is a standardized interface for enabling ergonomic and application-specific queries to on-chain state. Move view functions are developer defined on-chain read API that can be easily queried off-chain without requiring transaction signing or state mutation. In a Move module, the view function is annotated with the `#[view]` attribute.  
The attribute marks the function as **read-only**: it may inspect on-chain state but cannot modify it. IOTA enforces this guarantee at both compile time and verifier time, and records it on-chain, so that off-chain consumers can rely on a function being a view without trusting its author.

> 
> ## Why view functions exist
> Reading data from a Move package has always been possible by simulating a call, but nothing about a function's declaration told a client whether calling it was safe to treat as a pure read. Moreover, one must write a significant amount of custom client-side logic to inspect on-chain data. This often involves multiple layers of fetching object IDs, deserializing raw bytes, and navigating dynamic fields just to reach an information stored within a Move object. 

Reviewer 2: ```suggestion
## Why view functions exist

Reading data from a Move package has always been possible by simulating a call, but nothing about a function's declaration told a client whether calling it was safe to treat as a pure read. Moreover, one must write a significant amount of custom client-side logic to inspect on-chain data. This often involves multiple layers of fetching object IDs, deserializing raw bytes, and navigating dynamic fields just to reach an information stored within a Move object. 
```

---
description: 'What view functions are in Move on IOTA, the read-only guarantees they provide, and how those guarantees are enforced and recorded on-chain.'
tags:
  - explanation
  - move-sc
  - move-vm
---

# View Functions

> A _ Move View Function_ is a standardized interface for enabling ergonomic and application-specific queries to on-chain state. Move view functions are developer defined on-chain read API that can be easily queried off-chain without requiring transaction signing or state mutation. In a Move module, the view function is annotated with the `#[view]` attribute.  

Author: ```suggestion
A _Move View Function_ is a standardized interface for enabling ergonomic and application-specific queries to on-chain state. Move view functions are developer defined on-chain read API that can be easily queried off-chain without requiring transaction signing or state mutation. In a Move module, the view function is annotated with the `#[view]` attribute.  
```

## Why view functions exist

Reading data from a Move package has always been possible by simulating a call, but nothing about a function's declaration told a client whether calling it was safe to treat as a pure read. Moreover, one must write a significant amount of custom client-side logic to inspect on-chain data. This often involves multiple layers of fetching object IDs, deserializing raw bytes, and navigating dynamic fields just to reach an information stored within a Move object. 

The Move View Function closes that gap:

- **Package authors** express intent directly in the source, and the compiler holds them to it.
- **Client and SDK developers** can discover a package's view functions and call them knowing the call has no side effects.
- **Tooling** (explorers, indexers, wallets) can surface view functions as a safe, queryable read API for a package.
- **Extends current RPC** by enabling easy access to data structures using custom fields or dynamic fields, e.g., Bag and Table.

> - ** Returned results** are resolved, i.e., the Move types are deserialized.

Author: ```suggestion
- **Returned results** are resolved, i.e., the Move types are deserialized.
```

Given that a view function has no side effects, it can be called off-chain over JSON-RPC or GraphQL, or through an SDK, without submitting a transaction. A view call needs **no gas coin** and **no signature**. The returned values are deserialized to spare callers from decoding raw BCS bytes. See [Call a View Function](../how-tos/call-a-view-function.mdx).

The motivation, design, and on-chain encoding are specified in [IIP-0005](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0005/iip-0005.md) and [IIP-0011](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0011/IIP-0011.md).

## What is a `#[view]` function

A `#[view]` function cannot take a mutable reference (through which it could mutate state), cannot consume objects by value (which would let it destroy or transfer them), and must return a value. For the exact rules, see the [view functions reference](../../references/move/view-functions.mdx).

This is an example of view function that reads a `Bid` value from an `Auction` object:

> ```move file=<rootDir>/examples/move/view_functions/sources/auction.move#L83-L97

Reviewer: ```suggestion
```move file=<rootDir>/examples/move/view_functions/sources/auction.move#L93-L97
```
